### PR TITLE
Rule lifting rewrite

### PR DESCRIPTION
**Pull request by ajpal**
One regression test is failing due to not being able to derive `(xor ?a (or ?b ?a)) <=> (and ?b (not ?a))`
Still not exactly sure how/why we're losing that proving power. I'm looking into it, but I feel like it's still worth starting code review now

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -891,290 +869,166 @@ impl<L: SynthLanguage> Synthesizer<L> {
...
+        );
 
+        let mut candidates: EqualityMap<L> = EqualityMap::default();
         for ids in found_unions.values() {
-            for win in ids.windows(2) {
-                if self.params.keep_all
-                    || (self.egraph[win[0]].nodes.iter().any(L::is_extractable)
-                        && self.egraph[win[1]].nodes.iter().any(L::is_extractable))
-                {
+            for id1 in ids.clone() {
```
Comment (bksaiki) [inline src/lib.rs:980]: Is there a reason for this method of iteration rather than a sliding window of 2?

File: src/lib.rs [hunk 2]
```diff
@@ -891,290 +869,166 @@ impl<L: SynthLanguage> Synthesizer<L> {
...
                                 }
                             }
                         }
                     }
                 }
-
-                self.egraph.union(win[0], win[1]);
             }
         }
+        candidates.retain(|_, v| L::is_allowed_rewrite(&v.lhs, &v.rhs));
```
Comment (bksaiki) [inline src/lib.rs:1026]: So much cleaner!

**Review by bksaiki — COMMENTED** [review]
Looks much better. Possible things that need looking into
  (1) strange loss of derivability for one test
  ~~(2) effects on trig domain since the scheduler was changed
      (wouldn't be surprised if doesn't come back)~~
 (2) effects on rule lifting domains since the scheduler was changed.
    It seems the bv2 from bool domain went from 2 sec -> 87 sec in CI.
    I think it's safe to say trig won't come back.
    Scheduling is still a problem.

File: src/lib.rs [hunk 3]
```diff
@@ -448,25 +448,11 @@ impl<L: SynthLanguage> Synthesizer<L> {
...
-                } else {
-                    Ok(())
-                }
-            })
-            .with_node_limit(self.params.eqsat_node_limit * 2)
+            .with_node_limit(self.params.eqsat_node_limit)
             .with_iter_limit(self.params.eqsat_iter_limit)
             .with_time_limit(Duration::from_secs(self.params.eqsat_time_limit))
-            .with_scheduler(scheduler)
+            .with_scheduler(SimpleScheduler)
```
[the remark this answers is not in the document]
Comment (bksaiki) [reply src/lib.rs:455]: See note below on effects in rule lifting domain. Moving to `SimpleScheduler` is questionable.

**Comment by bksaiki** [comment]
Further investigation suggests that scheduling is not the problem (91 seconds for Backoff, 95 seconds for SimpleScheduler). Maybe the new unioning scheme?

File: src/lib.rs (hunk 1 above)
Comment (bksaiki) [reply src/lib.rs:980]: I think a sliding window is better. This is causing the performance blow up. Is there some factor I'm missing?

Comment (ajpal) [reply src/lib.rs:980]: There was a discussion about this in [this slack thread](https://uw-cse.slack.com/archives/C01749FQE4U/p1662153465032329)
the gist is- by using a sliding window and unioning as you go, you only see a subset of potential candidates in an arbitrary/unpredictable order

File: src/bin/trig.rs [hunk 4]
```diff
@@ -232,44 +232,11 @@ impl SynthLanguage for Math {
...
-                // Interesting complex nodes
-                // | Math::Cis(_)
-                // | Math::Imag
-
-                // Reciprocal trig
-                | Math::Csc(_)
-                | Math::Sec(_)
-                | Math::Cot(_)
-        )
+        !matches!(self, Math::Imag | Math::Cis(_),)
```
Comment (chandrakananandi) [inline src/bin/trig.rs:235]: great!

File: src/lib.rs [hunk 5]
```diff
@@ -891,290 +869,176 @@ impl<L: SynthLanguage> Synthesizer<L> {
...
+            old_eqs,
+            smt_unknown: self.smt_unknown,
+        }
+    }
 
-        let mut old_allowed_eqs: EqualityMap<L> = EqualityMap::default();
-        let mut old_forbidden_eqs: EqualityMap<L> = EqualityMap::default();
-        for (name, eq) in self.old_eqs.clone() {
+    fn run_chunk_rule_lifting(&mut self, chunk: &[RecExpr<L>]) {
+        log::info!("Adding {} terms to egraph", chunk.len());
```
Comment (chandrakananandi) [inline src/lib.rs:929]: is this `step 1. Add workload terms to egraph ` from the ruler doc? If so, can we add a comment?

**Review by chandrakananandi — APPROVED** [review]
I took a quick look at the PR and matched it with the algorithm we wrote in the ruler doc. It seems great to me! Left a suggestion for a comment but rest is good :shipit:

**Pull request merged**